spyder/utils/misc.py

```diff
@@ -323,3 +326,44 @@ def check_connection_port(address, port):
         return False
     finally:
         s.close()
+
+
+def get_pyenv_path(name):
+    """Return the complete path of the pyenv."""
+    home = get_home_dir()
+    if WINDOWS:
+        path = osp.join(
+            home, '.pyenv', 'pyenv-win', 'versions', name, 'python')
+    elif name == '':
+        path = osp.join(home, '.pyenv', 'shims', 'python')
+    else:
+        path = osp.join(home, '.pyenv', 'versions', name, 'bin', 'python')
+    return path
+
+
+def get_list_pyenv_envs():
+    """Return the list of all pyenv envs found in the system."""
+    try:
+        out, err = subprocess.Popen(
+            ['pyenv', 'versions', '--bare', '--skip-aliases'],

```

Comment:
@steff456 @ccordoba12, I wonder if we could simplify this by instead enhancing `is_program_installed` (called by `find_program`). What if `is_program_installed` checked that certain paths were already in `PATH` and _temporarily_ added them if not present, based on platform. This could do 2 things
1. Other developers could use `find_program` without having to consider the macOS application case.
1. Even outside of the macOS application, when searching for a program on the system, we don't have to assume that user profiles have been sourced (e.g. ~/.bash_Profile, etc)

What do you think?

`spyder.utils.programs.py`
```python
def is_program_installed(basename):
    """
    Return program absolute path if installed in PATH.

    Otherwise, return None

    On macOS systems, a .app is considered installed if
    it exists.
    """
    if (sys.platform == 'darwin' and basename.endswith('.app') and
            osp.exists(basename)):
        return basename

    home = get_home_dir()
    req_paths = []
    if sys.platform == 'darwin':
        req_paths.extend([
            osp.join('/usr', 'local', 'bin'),
            osp.join(home, 'opt', 'anaconda3', 'condabin'),
            osp.join(home, 'opt', 'miniconda3', 'condabin'),
            osp.join('/opt', 'anaconda3', 'condabin'),
            osp.join('/opt', 'miniconda3', 'condabin')
        ])
    elif sys.platform.startswith('linux'):
        pass  # Set required linux paths
    elif os.name == 'nt':
        pass  # Set required Windows paths

    old_path = os.environ['PATH']
    for path in req_paths:
        if path not in old_path:
            os.environ['PATH'] += os.pathsep + path

    abspath = None
    for path in os.environ["PATH"].split(os.pathsep):
        abspath = osp.join(path, basename)
        if osp.isfile(abspath):
            break
    os.environ['PATH'] = old_path

    return abspath
```

Reply:
I think it is a great idea! In this way we can guarantee this type of tooling to work inside the macOS application. @ccordoba12 do you think we need to add this in a different PR or in this one?

Reply:
I think we should leave this for another PR. But I agree it's a really good idea.

Reply:
Also, it'd great if you could give us a hand with that @mrclary.